File: src/jsonrpc/eth/getLogs/eth_getLogs.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Returns an array of all logs matching the specified filter.
//
// Example:
// Filter: ...
// Result: ...
//
// Implements the eth_getLogs JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_getLogs"

// Params represents the parameters for eth_getLogs
type Params struct {
	// filter
	Filter types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
		p.Filter,
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 1 {
		return fmt.Errorf("expected 1 parameters, got %d", len(arr))
	}

	if err := json.Unmarshal(arr[0], &p.Filter); err != nil {
		return fmt.Errorf("parameter 0 (Filter): %w", err)
	}

	return nil
}

// Result represents the result for eth_getLogs
//
// Filter results
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}

File: src/jsonrpc/eth/getLogs/eth_getLogs.js
/**
 * @fileoverview eth_getLogs JSON-RPC method
 */

/**
 * @typedef {import('../../../primitives/Address/AddressType.js').AddressType} Address
 * @typedef {import('../../index.js').Hash} Hash
 * @typedef {import('../../index.js').Quantity} Quantity
 * @typedef {import('../../index.js').BlockTag} BlockTag
 * @typedef {import('../../index.js').BlockSpec} BlockSpec
 * @typedef {import('../../../provider/types.js').RequestArguments} RequestArguments
 */

/**
 * Returns an array of all logs matching the specified filter.
 *
 * @example
 * Filter: ...
 * Result: ...
 *
 * Implements the `eth_getLogs` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_getLogs";
/**
 * Result for `eth_getLogs`
 *
 * Filter results
 *
 * @typedef {Quantity} Result
 */

/**
 * Creates a eth_getLogs JSON-RPC request
 *
 * @param {any} address
 * @returns {RequestArguments}
 */
export function GetLogsRequest(address) {
	return { method, params: [address] };
}
